File: app/src/main/java/com/jsoft/jcomic/FullscreenActivity.java
package com.jsoft.jcomic;

import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.KeyEvent;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.SeekBar;

import com.google.gson.Gson;
import com.jsoft.jcomic.adapter.FullScreenImageAdapter;
import com.jsoft.jcomic.helper.BookDTO;
import com.jsoft.jcomic.helper.BookmarkDb;
import com.jsoft.jcomic.helper.ComicsViewPager;
import com.jsoft.jcomic.helper.EpisodeDTO;
import com.jsoft.jcomic.helper.Utils;
import com.jsoft.jcomic.praser.CartoonMadEpisodeParser;
import com.jsoft.jcomic.praser.ComicVIPEpisodeParser;
import com.jsoft.jcomic.praser.DM5EpisodeParser;
import com.jsoft.jcomic.praser.EpisodeParser;
import com.jsoft.jcomic.praser.EpisodeParserListener;

import java.io.File;
import java.io.FileReader;
import java.util.ArrayList;

public class FullscreenActivity extends AppCompatActivity implements
        SeekBar.OnSeekBarChangeListener, EpisodeParserListener {
    private int pageTurn;
    private boolean gotoLastPage;
    private BookDTO book;
    private int currEpisode;
    private BookmarkDb bookmarkDb;
    private ComicsViewPager pager;
    private LinearLayout seekBarLayout;
    private SeekBar seekBar;
    private Runnable hideSeekBarTask;
    private boolean isSeeking = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        bookmarkDb = new BookmarkDb(this);
        Intent i = getIntent();
        currEpisode = i.getIntExtra("position", 0);
        Bundle b = this.getIntent().getExtras();
        if(b!=null)
            book = (BookDTO) b.getSerializable("book");
        else {
            book = new BookDTO("");
            book.setEpisodes(new ArrayList<EpisodeDTO>());
        }
        //Log.d("jComic", "EpisodeUrl: " + episode.getEpisodeUrl());
        EpisodeDTO episode = book.getEpisodes().get(currEpisode);
        if (Utils.isInternetAvailable()) {
            EpisodeParser.parseEpisode(episode, this);
        } else {
            try {
                File episodeFile = new File(Utils.getEpisodeFile(book, episode), "episode.json");
                if (episodeFile.exists()) {
                    Gson gson = new Gson();
                    EpisodeDTO savedEpisode = gson.fromJson(new FileReader(episodeFile.getAbsolutePath()), EpisodeDTO.class);
                    onEpisodeFetched(savedEpisode);
                } else {
                    onEpisodeFetched(episode);
                }
            } catch (Exception e) {
                Log.e("jComics", "Error caught in reading saved episode.", e);
            }
        }
    }

    public void onEpisodeFetched(EpisodeDTO episode) {
        setContentView(R.layout.activity_fullscreen);
        this.seekBarLayout = (LinearLayout) findViewById(R.id.seekbar_layout);
        this.seekBarLayout.setVisibility(View.GONE);
        seekBar = (SeekBar) findViewById(R.id.seekBar);
        seekBar.setMax(episode.getPageCount() - 1);
        seekBar.setOnSeekBarChangeListener(this);
        pager = (ComicsViewPager) findViewById(R.id.pager);
        pager.setActivity(this);
        if (pager.getAdapter() == null) {
            FullScreenImageAdapter adapter = new FullScreenImageAdapter(FullscreenActivity.this, pager, episode, book);
            pager.setAdapter(adapter);
        } else {
            ((FullScreenImageAdapter)pager.getAdapter()).setEpisode(book.getEpisodes().get(currEpisode));
        }
        int currentPage = 0;
        String lastEpisode = bookmarkDb.getLastEpisode(book);
        if (book.getEpisodes().get(currEpisode).getEpisodeTitle().equals(lastEpisode)) {
            currentPage = bookmarkDb.getLastEpisodePage(book);
        }
        if (pageTurn == -1 && gotoLastPage) {
            currentPage = episode.getPageCount()-1;
        }
        pager.setOffscreenPageLimit(2);
        if (!bookmarkDb.bookInDb(book)) {
            bookmarkDb.insertBookIntoDb(book);
        }
        switchPageNum(currentPage);
    }

    public void showPageBar() {
        this.seekBarLayout.setVisibility(View.VISIBLE);
        final View view = this.seekBarLayout;
        view.removeCallbacks(hideSeekBarTask);
        hideSeekBarTask = new Runnable() {
            public void run() {
                view.setVisibility(View.GONE);
            }
        };
        view.postDelayed(hideSeekBarTask, 5000);
    }

    @Override
    public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
        if (!isSeeking) {
            switchPageNum(progress);
        }
    }

    @Override
    public void onStartTrackingTouch(SeekBar seekBar) {
        final View view = this.seekBarLayout;
        view.removeCallbacks(hideSeekBarTask);
        isSeeking = true;
    }

    @Override
    public void onStopTrackingTouch(SeekBar seekBar) {
        isSeeking = false;
        switchPageNum(seekBar.getProgress());
        final View view = this.seekBarLayout;
        hideSeekBarTask = new Runnable() {
            public void run() {
                view.setVisibility(View.GONE);
            }
        };
        view.postDelayed(hideSeekBarTask, 5000);
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if ((keyCode == KeyEvent.KEYCODE_VOLUME_DOWN)){
            pager.turnNext();
            return true;
        } else if ((keyCode == KeyEvent.KEYCODE_VOLUME_UP)){
            pager.turnPrev();
            return true;
        }
        return false;
    }

    public void episodeSwitch(int pageTurn) {
        episodeSwitch(pageTurn, true);
    }

    public void episodeSwitch(int pageTurn, boolean gotoLastPage) {
        currEpisode = currEpisode - pageTurn;
        this.pageTurn = pageTurn;
        this.gotoLastPage = gotoLastPage;
        if (currEpisode >=0 && currEpisode < book.getEpisodes().size()) {
            if (book.getEpisodes().get(currEpisode).getPageCount() > 0) {
                onEpisodeFetched(book.getEpisodes().get(currEpisode));
            } else {
                if (book.getEpisodes().get(currEpisode).getEpisodeUrl().contains("comicbus")) {
                    new ComicVIPEpisodeParser(book.getEpisodes().get(currEpisode), this);
                } else if (book.getEpisodes().get(currEpisode).getEpisodeUrl().contains("cartoonmad")) {
                    new CartoonMadEpisodeParser(book.getEpisodes().get(currEpisode), this);
                } else if (book.getEpisodes().get(currEpisode).getEpisodeUrl().contains("dm5.com")) {
                    new DM5EpisodeParser(book.getEpisodes().get(currEpisode), this);
                }
            }
        } else {
            this.finish();
        }
    }

    public void switchPageNum(int pageNum) {
        pager.setCurrentItem(pageNum);
        seekBar.setProgress(pageNum);
        bookmarkDb.updateLastRead(book, currEpisode, pageNum);
    }

    public void goPrev(View view) {
        episodeSwitch(-1, false);
    }

    public void goNext(View view) {
        episodeSwitch(1);
    }

}

File: app/src/main/java/com/jsoft/jcomic/FullscreenActivity.kt
package com.jsoft.jcomic

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.view.KeyEvent
import android.view.View
import android.widget.LinearLayout
import android.widget.SeekBar
import com.google.gson.Gson
import com.jsoft.jcomic.adapter.FullScreenImageAdapter
import com.jsoft.jcomic.helper.*
import com.jsoft.jcomic.praser.*
import java.io.File
import java.io.FileReader

class FullscreenActivity : AppCompatActivity(), SeekBar.OnSeekBarChangeListener, EpisodeParserListener {
    private var pageTurn: Int = 0
    private var gotoLastPage: Boolean = false
    private var book: BookDTO? = null
    private var currEpisode: Int = 0
    private var bookmarkDb: BookmarkDb? = null
    private var pager: ComicsViewPager? = null
    private var seekBarLayout: LinearLayout? = null
    private var seekBar: SeekBar? = null
    private var hideSeekBarTask: Runnable? = null
    private var isSeeking = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        bookmarkDb = BookmarkDb(this)
        val i = intent
        currEpisode = i.getIntExtra("position", 0)
        val b = this.intent.extras
        book = if (b != null)
            b.getSerializable("book") as BookDTO
        else {
            BookDTO("")
        }
        //Log.d("jComic", "EpisodeUrl: " + episode.getEpisodeUrl());
        val episode = book!!.episodes[currEpisode]

        try {
            val episodeFile = File(Utils.getEpisodeFile(book!!, episode), "episode.json")
            if (episodeFile.exists()) {
                val gson = Gson()
                val savedEpisode = gson.fromJson(FileReader(episodeFile.absolutePath), EpisodeDTO::class.java)
                onEpisodeFetched(savedEpisode)
            } else if (Utils.isInternetAvailable) {
                EpisodeParser.parseEpisode(episode, this)
            } else {
                onEpisodeFetched(episode)
            }
        } catch (e: Exception) {
            Log.e("jComics", "Error caught in reading saved episode.", e)
        }

    }

    override fun onWindowFocusChanged(hasFocus: Boolean) {
        super.onWindowFocusChanged(hasFocus)
        if (hasFocus) hideSystemUI()
    }

    private fun hideSystemUI() {
        // Enables regular immersive mode.
        // For "lean back" mode, remove SYSTEM_UI_FLAG_IMMERSIVE.
        // Or for "sticky immersive," replace it with SYSTEM_UI_FLAG_IMMERSIVE_STICKY
        window.decorView.systemUiVisibility = (View.SYSTEM_UI_FLAG_IMMERSIVE
                // Set the content to appear under the system bars so that the
                // content doesn't resize when the system bars hide and show.
                or View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                or View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                // Hide the nav bar and status bar
                or View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
                or View.SYSTEM_UI_FLAG_FULLSCREEN)
    }

    // Shows the system bars by removing all the flags
// except for the ones that make the content appear under the system bars.
    private fun showSystemUI() {
        window.decorView.systemUiVisibility = (View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                or View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN)
    }

    override fun onEpisodeFetched(episode: EpisodeDTO) {
        setContentView(R.layout.activity_fullscreen)
        this.seekBarLayout = findViewById(R.id.seekbar_layout)
        this.seekBarLayout!!.visibility = View.GONE
        seekBar = findViewById(R.id.seekBar)
        seekBar!!.max = episode.pageCount - 1
        seekBar!!.setOnSeekBarChangeListener(this)
        pager = findViewById(R.id.pager)
        pager!!.setActivity(this)
        if (pager!!.adapter == null) {
            val adapter = FullScreenImageAdapter(this@FullscreenActivity, pager!!, episode, book!!)
            pager!!.adapter = adapter
        } else {
            (pager!!.adapter as FullScreenImageAdapter).episode = book!!.episodes[currEpisode]
        }
        var currentPage = 0
        val lastEpisode = bookmarkDb!!.getLastEpisode(book!!)
        if (book!!.episodes[currEpisode].episodeTitle == lastEpisode) {
            currentPage = bookmarkDb!!.getLastEpisodePage(book!!)
        }
        if (pageTurn == -1 && gotoLastPage) {
            currentPage = episode.pageCount - 1
        }
        pager!!.offscreenPageLimit = 2
        if (!bookmarkDb!!.bookInDb(book!!)) {
            bookmarkDb!!.insertBookIntoDb(book!!)
        }
        switchPageNum(currentPage)
    }

    fun showPageBar() {
        this.seekBarLayout!!.visibility = View.VISIBLE
        val view = this.seekBarLayout
        view!!.removeCallbacks(hideSeekBarTask)
        hideSeekBarTask = Runnable { view.visibility = View.GONE }
        view.postDelayed(hideSeekBarTask, 5000)
    }

    override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
        if (!isSeeking) {
            switchPageNum(progress)
        }
    }

    override fun onStartTrackingTouch(seekBar: SeekBar) {
        val view = this.seekBarLayout
        view!!.removeCallbacks(hideSeekBarTask)
        isSeeking = true
    }

    override fun onStopTrackingTouch(seekBar: SeekBar) {
        isSeeking = false
        switchPageNum(seekBar.progress)
        val view = this.seekBarLayout
        hideSeekBarTask = Runnable { view!!.visibility = View.GONE }
        view!!.postDelayed(hideSeekBarTask, 5000)
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        if (keyCode == KeyEvent.KEYCODE_VOLUME_DOWN) {
            pager!!.turnNext()
            return true
        } else if (keyCode == KeyEvent.KEYCODE_VOLUME_UP) {
            pager!!.turnPrev()
            return true
        }
        return super.onKeyDown(keyCode, event)
    }

    @JvmOverloads
    fun episodeSwitch(pageTurn: Int, gotoLastPage: Boolean = true) {
        currEpisode -= pageTurn
        this.pageTurn = pageTurn
        this.gotoLastPage = gotoLastPage
        if (currEpisode >= 0 && currEpisode < book!!.episodes.size) {
            if (book!!.episodes[currEpisode].pageCount > 0) {
                onEpisodeFetched(book!!.episodes[currEpisode])
            } else {
                if (book!!.episodes[currEpisode].episodeUrl!!.contains("comicbus")) {
                    ComicVIPEpisodeParser(book!!.episodes[currEpisode], this)
                } else if (book!!.episodes[currEpisode].episodeUrl!!.contains("cartoonmad")) {
                    CartoonMadEpisodeParser(book!!.episodes[currEpisode], this)
                } else if (book!!.episodes[currEpisode].episodeUrl!!.contains("dm5.com")) {
                    DM5EpisodeParser(book!!.episodes[currEpisode], this)
                }
            }
        } else {
            this.finish()
        }
    }

    fun switchPageNum(pageNum: Int) {
        pager!!.currentItem = pageNum
        seekBar!!.progress = pageNum
        bookmarkDb!!.updateLastRead(book!!, currEpisode, pageNum)
    }

    fun goPrev(view: View) {
        episodeSwitch(-1, false)
    }

    fun goNext(view: View) {
        episodeSwitch(1)
    }

}
